File: ios/RNLWeChatAPIObjectConvert.m
#import <React/RCTConvert.h>
#import <React/RCTUtils.h>

#import "RNLWeChatAPIObjectConvert.h"

const NSUInteger kWeChatThumbSizeLimit = 65536; // 64 KB

NSData *__nullable RNLParseFileStr(NSString *str) {
    if (!str) {
        return nil;
    }
    if ([str hasPrefix:@"~"] ||
        [str hasPrefix:@"/"] ||
        [str hasPrefix:@"file"] ||
        [str hasPrefix:@"http"]) {
        NSURL *url = [RCTConvert NSURL:str];
        return [NSData dataWithContentsOfURL:url];
    } else if ([str hasPrefix:@"data:"]) {
        str = [str substringFromIndex:[str rangeOfString:@","].location + 1];
    }
    return [[NSData alloc] initWithBase64EncodedString:str
                                               options:NSDataBase64DecodingIgnoreUnknownCharacters];
}

@implementation RNLWeChatAPIObjectConvert

+ (NSDictionary *)CardItemFromWXO:(WXCardItem *)aWXCardItem
{
    NSMutableDictionary *card = [NSMutableDictionary new];
    card[@"id"] = aWXCardItem.cardId;
    card[@"ext"] = aWXCardItem.extMsg;
    card[@"state"] = [NSNumber numberWithUnsignedInteger:aWXCardItem.cardState];
    card[@"encryptCode"] = aWXCardItem.encryptCode;
    card[@"appID"] = aWXCardItem.appID;
    return card;
}

+ (NSDictionary *)InvoiceItemFromWXO:(WXInvoiceItem *)aWXInvoiceItem
{
    NSMutableDictionary *invoice = [NSMutableDictionary new];
    invoice[@"id"] = aWXInvoiceItem.cardId;
    invoice[@"ext"] = aWXInvoiceItem.extMsg;
    invoice[@"state"] = [NSNumber numberWithUnsignedInteger:aWXInvoiceItem.cardState];
    invoice[@"encryptCode"] = aWXInvoiceItem.encryptCode;
    invoice[@"appID"] = aWXInvoiceItem.appID;
    return invoice;
}

+ (NSDictionary *)MediaMessageFromWXO:(WXMediaMessage *)aMediaMessage
{
    NSMutableDictionary *message = [NSMutableDictionary new];
    message[@"messageTitle"] = NSStringNonNull(aMediaMessage.title);
    message[@"messageDesc"] = NSStringNonNull(aMediaMessage.description);
    NSURL *thumb = RCTDataURL(@"image/png", aMediaMessage.thumbData);
    message[@"messageThumb"] = thumb ? [thumb absoluteString] : @"";
    message[@"messageTag"] = NSStringNonNull(aMediaMessage.mediaTagName);
    message[@"messageExt"] = NSStringNonNull(aMediaMessage.messageExt);
    message[@"messageAction"] = NSStringNonNull(aMediaMessage.messageAction);
    
    id mediaObject = aMediaMessage.mediaObject;

    if (mediaObject != nil) {
        if ([mediaObject isKindOfClass:[WXTextObject class]]) {
            [message addEntriesFromDictionary:
             [self TextFromWXO:(WXTextObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXImageObject class]]) {
            [message addEntriesFromDictionary:
             [self ImageFromWXO:(WXImageObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXMusicObject class]]) {
            [message addEntriesFromDictionary:
             [self MusicFromWXO:(WXMusicObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXVideoObject class]]) {
            [message addEntriesFromDictionary:
             [self VideoFromWXO:(WXVideoObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXFileObject class]]) {
            [message addEntriesFromDictionary:
             [self FileFromWXO:(WXFileObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXWebpageObject class]]) {
            [message addEntriesFromDictionary:
             [self WebpageFromWXO:(WXWebpageObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXMiniProgramObject class]]) {
            [message addEntriesFromDictionary:
             [self MiniProgramFromWXO:(WXMiniProgramObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXAppExtendObject class]]) {
            [message addEntriesFromDictionary:
             [self AppExtendFromWXO:(WXAppExtendObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXEmoticonObject class]]) {
            [message addEntriesFromDictionary:
             [self EmoticonFromWXO:(WXEmoticonObject *)mediaObject]];

        } else if ([mediaObject isKindOfClass:[WXLocationObject class]]) {
            [message addEntriesFromDictionary:
             [self LocationFromWXO:(WXLocationObject *)mediaObject]];
        }
    }

    return message;
}

+ (WXMediaMessage *)MediaMessageToWXO:(NSDictionary *)aMessageData
{
    NSInteger messageType = [RCTConvert NSInteger:aMessageData[@"messageType"]];

    switch (messageType) {
        case RNLWeChatTextMessage:
        {
            return nil;
        }
        case RNLWeChatImageMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.mediaObject = [self ImageToWXO:aMessageData];
            // 发送到收藏时, 预览显示用此数据
            if ([[((WXImageObject *) message.mediaObject) imageData] length] < kWeChatThumbSizeLimit) {
                [message setThumbData:((WXImageObject *)message.mediaObject).imageData];
            }
            return message;
        }
        case RNLWeChatMusicMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.title = [RCTConvert NSString:aMessageData[@"messageTitle"]];
            message.description = [RCTConvert NSString:aMessageData[@"messageDesc"]];
            [message setThumbData:RNLParseFileStr(aMessageData[@"messageThumb"])];
            message.mediaObject = [self MusicToWXO:aMessageData];
            return message;
        }
        case RNLWeChatVideoMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.title = [RCTConvert NSString:aMessageData[@"messageTitle"]];
            message.description = [RCTConvert NSString:aMessageData[@"messageDesc"]];
            [message setThumbData:RNLParseFileStr(aMessageData[@"messageThumb"])];
            message.mediaObject = [self VideoToWXO:aMessageData];
            return message;
        }
        case RNLWeChatFileMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.title = [RCTConvert NSString:aMessageData[@"messageTitle"]];
            message.mediaObject = [self FileToWXO:aMessageData];
            return message;
        }
        case RNLWeChatWebpageMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.title = [RCTConvert NSString:aMessageData[@"messageTitle"]];
            message.description = [RCTConvert NSString:aMessageData[@"messageDesc"]];
            [message setThumbData:RNLParseFileStr(aMessageData[@"messageThumb"])];
            message.mediaObject = [self WebpageToWXO:aMessageData];
            return message;
        }
        case RNLWeChatMiniProgramMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.title = [RCTConvert NSString:aMessageData[@"messageTitle"]];
            message.description = [RCTConvert NSString:aMessageData[@"messageDesc"]];
            [message setThumbData:RNLParseFileStr(aMessageData[@"messageThumb"])];
            message.mediaObject = [self MiniProgramToWXO:aMessageData];
            return message;
        }
        case RNLWeChatAppExtendMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.title = [RCTConvert NSString:aMessageData[@"messageTitle"]];
            message.description = [RCTConvert NSString:aMessageData[@"messageDesc"]];
            [message setThumbData:RNLParseFileStr(aMessageData[@"messageThumb"])];
            message.mediaTagName = [RCTConvert NSString:aMessageData[@"messageTag"]];
            message.messageExt = [RCTConvert NSString:aMessageData[@"messageExt"]];
            message.messageAction = [RCTConvert NSString:aMessageData[@"messageAction"]];
            message.mediaObject = [self AppExtendToWXO:aMessageData];
            return message;
        }
        case RNLWeChatEmoticonMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            [message setThumbData:RNLParseFileStr(aMessageData[@"image"])];
            message.mediaObject = [self EmoticonToWXO:aMessageData];
            return message;
        }
        case RNLWeChatLocationMessage:
        {
            WXMediaMessage *message = [WXMediaMessage message];
            message.title = [RCTConvert NSString:aMessageData[@"messageTitle"]];
            message.description = [RCTConvert NSString:aMessageData[@"messageDesc"]];
            [message setThumbData:RNLParseFileStr(aMessageData[@"messageThumb"])];
            message.mediaObject = [self LocationToWXO:aMessageData];
            return message;
        }
    }
    
    return nil;
}

+ (NSDictionary *)TextFromWXO:(WXTextObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"content"] = NSStringNonNull(anObject.contentText);
    return result;
}

+ (WXTextObject *)TextToWXO:(NSDictionary *)aData
{
    WXTextObject *object = [WXTextObject object];
    object.contentText = [RCTConvert NSString:aData[@"content"]];
    return object;
}

+ (NSDictionary *)ImageFromWXO:(WXImageObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    NSURL *image = RCTDataURL(@"image/png", anObject.imageData);
    result[@"image"] = image ? [image absoluteString] : @"";
    return result;
}

+ (WXImageObject *)ImageToWXO:(NSDictionary *)aData
{
    WXImageObject *object = [WXImageObject object];
    object.imageData = RNLParseFileStr(aData[@"image"]);
    return object;
}

+ (NSDictionary *)MusicFromWXO:(WXMusicObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"url"] = NSStringNonNull(anObject.musicUrl);
    result[@"lowBandURL"] = NSStringNonNull(anObject.musicLowBandUrl);
    result[@"dataURL"] = NSStringNonNull(anObject.musicDataUrl);
    result[@"lowBandDataURL"] = NSStringNonNull(anObject.musicLowBandDataUrl);
    return result;
}

+ (WXMusicObject *)MusicToWXO:(NSDictionary *)aData
{
    WXMusicObject *object = [WXMusicObject object];
    object.musicUrl = [RCTConvert NSString:aData[@"url"]];
    object.musicLowBandUrl = [RCTConvert NSString:aData[@"lowBandURL"]];
    object.musicDataUrl = [RCTConvert NSString:aData[@"dataURL"]];
    object.musicLowBandDataUrl = [RCTConvert NSString:aData[@"lowBandDataURL"]];
    return object;
}

+ (NSDictionary *)VideoFromWXO:(WXVideoObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"url"] = NSStringNonNull(anObject.videoUrl);
    result[@"lowBandURL"] = NSStringNonNull(anObject.videoLowBandUrl);
    return result;
}

+ (WXVideoObject *)VideoToWXO:(NSDictionary *)aData
{
    WXVideoObject *object = [WXVideoObject object];
    object.videoUrl = [RCTConvert NSString:aData[@"url"]];
    object.videoLowBandUrl = [RCTConvert NSString:aData[@"lowBandURL"]];
    return object;
}

+ (NSDictionary *)FileFromWXO:(WXFileObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"ext"] = NSStringNonNull(anObject.fileExtension);
    NSURL *file = RCTDataURL(anObject.fileExtension, anObject.fileData);
    result[@"file"] = file ? [file absoluteString] : @"";
    return result;
}

+ (WXFileObject *)FileToWXO:(NSDictionary *)aData
{
    WXFileObject *object = [WXFileObject object];
    object.fileExtension = [RCTConvert NSString:aData[@"ext"]];
    object.fileData = RNLParseFileStr(aData[@"file"]);
    return object;
}

+ (NSDictionary *)WebpageFromWXO:(WXWebpageObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"url"] = NSStringNonNull(anObject.webpageUrl);
    return result;
}

+ (WXWebpageObject *)WebpageToWXO:(NSDictionary *)aData
{
    WXWebpageObject *object = [WXWebpageObject object];
    object.webpageUrl = [RCTConvert NSString:aData[@"url"]];
    return object;
}

+ (NSDictionary *)MiniProgramFromWXO:(WXMiniProgramObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"type"] = [NSNumber numberWithUnsignedInteger:anObject.miniProgramType];
    result[@"username"] = NSStringNonNull(anObject.userName);
    result[@"path"] = NSStringNonNull(anObject.path);
    NSURL *hdImage = RCTDataURL(@"image/png", anObject.hdImageData);
    result[@"hdImage"] = hdImage ? [hdImage absoluteString] : @"";
    result[@"url"] = NSStringNonNull(anObject.webpageUrl);
    result[@"shareTicket"] = [NSNumber numberWithBool:anObject.withShareTicket];
    return result;
}

+ (WXMiniProgramObject *)MiniProgramToWXO:(NSDictionary *)aData
{
    WXMiniProgramObject *object = [WXMiniProgramObject object];
    object.miniProgramType = [RCTConvert NSUInteger:aData[@"type"]];
    object.userName = [RCTConvert NSString:aData[@"username"]];
    object.path = [RCTConvert NSString:aData[@"path"]];
    object.hdImageData = RNLParseFileStr(aData[@"hdImage"]);
    object.webpageUrl = [RCTConvert NSString:aData[@"url"]];
    object.withShareTicket = [RCTConvert BOOL:aData[@"shareTicket"]];
    return object;
}

+ (NSDictionary *)AppExtendFromWXO:(WXAppExtendObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"url"] = NSStringNonNull(anObject.url);
    result[@"ext"] = NSStringNonNull(anObject.extInfo);
    NSURL *file = RCTDataURL(@"", anObject.fileData);
    result[@"file"] = file ? [file absoluteString] : @"";
    return result;
}

+ (WXAppExtendObject *)AppExtendToWXO:(NSDictionary *)aData
{
    WXAppExtendObject *object = [WXAppExtendObject object];
    object.url = [RCTConvert NSString:aData[@"url"]];
    object.extInfo = [RCTConvert NSString:aData[@"ext"]];
    object.fileData = RNLParseFileStr(aData[@"file"]);
    return object;
}

+ (NSDictionary *)EmoticonFromWXO:(WXEmoticonObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    NSURL *image = RCTDataURL(@"image/png", anObject.emoticonData);
    result[@"image"] = image ? [image absoluteString] : @"";
    return result;
}

+ (WXEmoticonObject *)EmoticonToWXO:(NSDictionary *)aData
{
    WXEmoticonObject *object = [WXEmoticonObject object];
    object.emoticonData = RNLParseFileStr(aData[@"image"]);
    return object;
}

+ (NSDictionary *)LocationFromWXO:(WXLocationObject *)anObject
{
    NSMutableDictionary *result = [NSMutableDictionary new];
    result[@"lng"] = [NSNumber numberWithDouble:anObject.lng];
    result[@"lat"] = [NSNumber numberWithDouble:anObject.lat];
    return result;
}

+ (WXLocationObject *)LocationToWXO:(NSDictionary *)aData
{
    WXLocationObject *object = [WXLocationObject object];
    object.lng = [RCTConvert double:aData[@"lng"]];
    object.lat = [RCTConvert double:aData[@"lat"]];
    return object;
}

@end

File: android/src/main/java/com/rnlibrary/wechat/RNLWeChatAPIObjectConvert.java
package com.rnlibrary.wechat;

import android.content.Context;
import android.util.Base64;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.tencent.mm.opensdk.modelbiz.AddCardToWXCardPackage;
import com.tencent.mm.opensdk.modelmsg.WXAppExtendObject;
import com.tencent.mm.opensdk.modelmsg.WXEmojiObject;
import com.tencent.mm.opensdk.modelmsg.WXFileObject;
import com.tencent.mm.opensdk.modelmsg.WXImageObject;
import com.tencent.mm.opensdk.modelmsg.WXLocationObject;
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage;
import com.tencent.mm.opensdk.modelmsg.WXMiniProgramObject;
import com.tencent.mm.opensdk.modelmsg.WXMusicObject;
import com.tencent.mm.opensdk.modelmsg.WXTextObject;
import com.tencent.mm.opensdk.modelmsg.WXVideoObject;
import com.tencent.mm.opensdk.modelmsg.WXWebpageObject;

import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;

import javax.annotation.Nullable;

class RNLWeChatAPIObjectConvert {
    static WritableMap cardItemFromWXO(AddCardToWXCardPackage.WXCardItem cardItem) {
        WritableMap message = Arguments.createMap();
        message.putString("id", cardItem.cardId);
        message.putString("ext", cardItem.cardExtMsg);
        message.putInt("state", cardItem.cardState);
        return message;
    }

    static WritableMap mediaMessageFromWXO(WXMediaMessage mediaMessage) {
        WritableMap message = Arguments.createMap();
        message.putString("messageTitle", mediaMessage.title);
        message.putString("messageDesc", mediaMessage.description);
        try {
            message.putString("messageThumb",
                    Base64.encodeToString(mediaMessage.thumbData, Base64.DEFAULT));
        } catch (Exception ignored) {
        }
        message.putString("messageTag", mediaMessage.mediaTagName);
        message.putString("messageExt", mediaMessage.messageExt);
        message.putString("messageAction", mediaMessage.messageAction);

        WXMediaMessage.IMediaObject object = mediaMessage.mediaObject;

        if (object != null) {
            if (object instanceof WXTextObject) {
                message = textFromWXO(message, (WXTextObject) object);
            } else if (object instanceof WXImageObject) {
                message = imageFromWXO(message, (WXImageObject) object);
            } else if (object instanceof WXMusicObject) {
                message = musicFromWXO(message, (WXMusicObject) object);
            } else if (object instanceof WXVideoObject) {
                message = videoFromWXO(message, (WXVideoObject) object);
            } else if (object instanceof WXFileObject) {
                message = fileFromWXO(message, (WXFileObject) object);
            } else if (object instanceof WXWebpageObject) {
                message = webpageFromWXO(message, (WXWebpageObject) object);
            } else if (object instanceof WXMiniProgramObject) {
                message = miniProgramFromWXO(message, (WXMiniProgramObject) object);
            } else if (object instanceof WXAppExtendObject) {
                message = appExtendFromWXO(message, (WXAppExtendObject) object);
            } else if (object instanceof WXEmojiObject) {
                message = emoticonFromWXO(message, (WXEmojiObject) object);
            } else if (object instanceof WXLocationObject) {
                message = locationFromWXO(message, (WXLocationObject) object);
            }
        }

        return message;
    }

    static WXMediaMessage mediaMessageToWXO(Context context, ReadableMap data) throws Exception {
        WXMediaMessage message = new WXMediaMessage();

        int messageType = data.getInt("messageType");
        if (RNLWeChatMessageType.Text.getType() == messageType) {
            message.mediaObject = textToWXO(data);
            message.description = ((WXTextObject) message.mediaObject).text;
        } else if (RNLWeChatMessageType.Image.getType() == messageType) {
            message.mediaObject = imageToWXO(context, data);
        } else if (RNLWeChatMessageType.Music.getType() == messageType) {
            message.title = data.getString("messageTitle");
            message.description = data.getString("messageDesc");
            message.thumbData = parseFileStr(context, data.getString("messageThumb"));
            message.mediaObject = musicToWXO(data);
        } else if (RNLWeChatMessageType.Video.getType() == messageType) {
            message.title = data.getString("messageTitle");
            message.description = data.getString("messageDesc");
            message.thumbData = parseFileStr(context, data.getString("messageThumb"));
            message.mediaObject = videoToWXO(data);
        } else if (RNLWeChatMessageType.File.getType() == messageType) {
            message.title = data.getString("messageTitle");
            message.mediaObject = fileToWXO(context, data);
        } else if (RNLWeChatMessageType.Webpage.getType() == messageType) {
            message.title = data.getString("messageTitle");
            message.description = data.getString("messageDesc");
            message.thumbData = parseFileStr(context, data.getString("messageThumb"));
            message.mediaObject = webpageToWXO(data);
        } else if (RNLWeChatMessageType.MiniProgram.getType() == messageType) {
            message.title = data.getString("messageTitle");
            message.description = data.getString("messageDesc");
            message.thumbData = parseFileStr(context, data.getString("hdImage"));
            message.mediaObject = miniProgramToWXO(data);
        } else if (RNLWeChatMessageType.AppExtend.getType() == messageType) {
            message.title = data.getString("messageTitle");
            message.description = data.getString("messageDesc");
            message.thumbData = parseFileStr(context, data.getString("messageThumb"));
            message.mediaTagName = data.getString("messageTag");
            message.messageExt = data.getString("messageExt");
            message.messageAction = data.getString("messageAction");
            message.mediaObject = appExtendToWXO(context, data);
        } else if (RNLWeChatMessageType.Emoticon.getType() == messageType) {
            message.thumbData = parseFileStr(context, data.getString("image"));
            message.mediaObject = emoticonToWXO(context, data);
        } else if (RNLWeChatMessageType.Location.getType() == messageType) {
            message.title = data.getString("messageTitle");
            message.description = data.getString("messageDesc");
            message.thumbData = parseFileStr(context, data.getString("messageThumb"));
            message.mediaObject = locationToWXO(data);
        }

        return message;
    }

    static WritableMap textFromWXO(WritableMap message, WXTextObject object) {
        message.putString("content", object.text);
        return message;
    }

    static WXTextObject textToWXO(ReadableMap data) {
        WXTextObject object = new WXTextObject();
        object.text = data.getString("content");
        return object;
    }

    static WritableMap imageFromWXO(WritableMap message, WXImageObject object) {
        message.putString("path", object.imagePath);
        try {
            message.putString("data", Base64.encodeToString(object.imageData, Base64.DEFAULT));
        } catch (Exception ignored) {
        }
        return message;
    }

    static WXImageObject imageToWXO(Context context, ReadableMap data) throws Exception {
        WXImageObject object = new WXImageObject();
        object.imageData = parseFileStr(context, data.getString("image"));
        return object;
    }

    static WritableMap musicFromWXO(WritableMap message, WXMusicObject object) {
        message.putString("url", object.musicUrl);
        message.putString("lowBandURL", object.musicLowBandUrl);
        message.putString("dataURL", object.musicDataUrl);
        message.putString("lowBandDataURL", object.musicLowBandDataUrl);
        return message;
    }

    static WXMusicObject musicToWXO(ReadableMap data) {
        WXMusicObject object = new WXMusicObject();
        object.musicUrl = data.getString("url");
        object.musicLowBandUrl = data.getString("lowBandURL");
        object.musicDataUrl = data.getString("dataURL");
        object.musicLowBandDataUrl = data.getString("lowBandDataURL");
        return object;
    }

    static WritableMap videoFromWXO(WritableMap message, WXVideoObject object) {
        message.putString("url", object.videoUrl);
        message.putString("lowBandURL", object.videoLowBandUrl);
        return message;
    }

    static WXVideoObject videoToWXO(ReadableMap data) {
        WXVideoObject object = new WXVideoObject();
        object.videoUrl = data.getString("url");
        object.videoLowBandUrl = data.getString("lowBandURL");
        return object;
    }

    static WritableMap fileFromWXO(WritableMap message, WXFileObject object) {
        message.putString("path", object.filePath);
        try {
            message.putString("file", Base64.encodeToString(object.fileData, Base64.DEFAULT));
        } catch (Exception ignored) {
        }
        return message;
    }

    static WXFileObject fileToWXO(Context context, ReadableMap data) throws Exception {
        WXFileObject object = new WXFileObject();
        object.fileData = parseFileStr(context, data.getString("file"));
        return object;
    }

    static WritableMap webpageFromWXO(WritableMap message, WXWebpageObject object) {
        message.putString("url", object.webpageUrl);
        return message;
    }

    static WXWebpageObject webpageToWXO(ReadableMap data) {
        WXWebpageObject object = new WXWebpageObject();
        object.webpageUrl = data.getString("url");
        return object;
    }

    static WritableMap miniProgramFromWXO(WritableMap message, WXMiniProgramObject object) {
        message.putInt("type", object.miniprogramType);
        message.putString("username", object.userName);
        message.putString("path", object.path);
        message.putString("url", object.webpageUrl);
        message.putBoolean("shareTicket", object.withShareTicket);
        return message;
    }

    static WXMiniProgramObject miniProgramToWXO(ReadableMap data) {
        WXMiniProgramObject object = new WXMiniProgramObject();
        object.miniprogramType = data.getInt("type");
        object.userName = data.getString("username");
        object.path = data.getString("path");
        object.webpageUrl = data.getString("url");
        object.withShareTicket = data.getBoolean("shareTicket");
        return object;
    }

    static WritableMap appExtendFromWXO(WritableMap message, WXAppExtendObject object) {
        message.putString("path", object.filePath);
        try {
            message.putString("file", Base64.encodeToString(object.fileData, Base64.DEFAULT));
        } catch (Exception ignored) {
        }
        message.putString("ext", object.extInfo);
        return message;
    }

    static WXAppExtendObject appExtendToWXO(Context context, ReadableMap data) throws Exception {
        WXAppExtendObject object = new WXAppExtendObject();
        object.fileData = parseFileStr(context, data.getString("file"));
        object.extInfo = data.getString("ext");
        return object;
    }

    static WritableMap emoticonFromWXO(WritableMap message, WXEmojiObject object) {
        message.putString("path", object.emojiPath);
        try {
            message.putString("image", Base64.encodeToString(object.emojiData, Base64.DEFAULT));
        } catch (Exception ignored) {
        }
        return message;
    }

    static WXEmojiObject emoticonToWXO(Context context, ReadableMap data) throws Exception {
        WXEmojiObject object = new WXEmojiObject();
        object.emojiData = parseFileStr(context, data.getString("image"));
        return object;
    }

    static WritableMap locationFromWXO(WritableMap message, WXLocationObject object) {
        message.putDouble("lng", object.lng);
        message.putDouble("lat", object.lat);
        return message;
    }

    static WXLocationObject locationToWXO(ReadableMap data) {
        WXLocationObject object = new WXLocationObject();
        object.lng = data.getDouble("lng");
        object.lat = data.getDouble("lat");
        return object;
    }

    static byte[] parseFileStr(Context context, String str) throws Exception {
        if (str.startsWith("/")) {
            str = "file://" + str;
            return inputStreamToByteArray(new FileInputStream(str));
        } else if (str.startsWith("file")) {
            return inputStreamToByteArray(new FileInputStream(str));
        } else if (str.startsWith("http")) {
            URL url = new URL(str);
            URLConnection connection = url.openConnection();
            connection.connect();
            return inputStreamToByteArray(connection.getInputStream());
        } else if (str.startsWith("data:")) {
            str = str.substring(str.indexOf(",") + 1);
        }
        return Base64.decode(str, Base64.DEFAULT);
    }

    static byte[] inputStreamToByteArray(@Nullable InputStream iStream) throws Exception {
        if (iStream == null) {
            return null;
        }
        ByteArrayOutputStream oStream = null;
        try {
            oStream = new ByteArrayOutputStream();
            int c;
            while ((c = iStream.read()) != -1) {
                oStream.write(c);
            }
        } finally {
            iStream.close();
            if (oStream != null) {
                oStream.close();
            }
        }
        return oStream.toByteArray();
    }
}
